Added autoScrollToTop and async to DiffAdapter

RKuanysh opened the pull request:
Added autoScrollToTop and async in DiffAdapter.
Changed RxDiffAdapter to extend DiffAdapter.

RKuanysh committed:
Added abstract methods in AbsRecyclerDelegatesAdapter moving them from RxDiffAdapter

RKuanysh committed:
Added DefaultDiffAdapter that calculates diff on main thread

RKuanysh committed:
Removed old deprecated RxDiffAdapter constructor

RKuanysh committed:
Simplified constructor

funriser commented on rxdiffadapter/src/main/java/com/revolut/rxdiffadapter/RxDiffAdapter.kt:
```diff
@@ -47,7 +44,7 @@ open class RxDiffAdapter @Deprecated("Replace with constructor without delegates
     val async: Boolean = false,
     private val autoScrollToTop: Boolean = false,
```
Is there a reason to keep `autoScrollToTop` and `detectMoves` as properties? They can be declared as simple params instead

funriser replied:
Is there a reason to keep autoScrollToTop and detectMoves as properties? They can be declared as simple params instead

funriser commented on delegates/src/main/java/com/revolut/recyclerkit/delegates/AbsRecyclerDelegatesAdapter.kt:
```diff
@@ -26,6 +26,8 @@ abstract class AbsRecyclerDelegatesAdapter(
     val delegatesManager: DelegatesManager = DelegatesManager()
 ) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
 
+    abstract val items: List<ListItem>
```
Let's avoid adding new abstract methods to the existing base class. Otherwise it will be a breaking API change which can't fit to the minor upgrade

RKuanysh replied:
removed

funriser commented on delegates/src/main/java/com/revolut/recyclerkit/delegates/DefaultDiffAdapter.kt:
```diff
@@ -0,0 +1,130 @@
+* @param detectMoves – true if DiffUtil should try to detect moved items, false otherwise.
+*/
+
+open class DefaultDiffAdapter(
```
`DefaultDiffAdapter` introduces a conflict with `DiffAdapter`. The naming doesn't suggest what's the difference between them.
We can merge their functionality together or come up with the more straightforward naming

RKuanysh replied:
removed

RKuanysh committed:
Removed DefaultDiffAdapter. Added autoScrollToTop into DiffAdapter

RKuanysh committed:
Reverted AbsRecyclerDelegatesAdapter changes

RKuanysh committed:
Reverted rxdiffadapter version bump

RKuanysh committed:
Added async parameter to DiffAdapter. RxDiffAdapter extends DiffAdapter

funriser commented on delegates/src/main/java/com/revolut/recyclerkit/delegates/DiffAdapter.kt:
```diff
@@ -25,21 +31,53 @@ import androidx.recyclerview.widget.DiffUtil
+            }
+            .build()
     )
+    protected var recyclerView = WeakReference<RecyclerView>(null)
```
I would add `private set` to limit mutability for descendant classes

RKuanysh committed:
DiffAdapter with async and autoScrollToTop

funriser commented on delegates/src/main/java/com/revolut/recyclerkit/delegates/DiffAdapter.kt:
```diff
@@ -25,22 +28,113 @@ import androidx.recyclerview.widget.DiffUtil
+        fun setItems(items: List<ListItem>, recyclerView: RecyclerView)
+    }
+
+    private val differDelegate = if (async) AsyncDifferDelegate(this, autoScrollToTop) else SyncDifferDelegate(this, autoScrollToTop)
```
Nice approach with delegating differ logic to the separate abstraction
Let's reuse `SyncDifferDelegate` in `RxDiffAdapter`. Otherwise we have duplicated logic in the parent and child classes

funriser replied:
You can also use naming like AsyncDifferStrategy and SyncDifferStrategy because those delegates represent strategy pattern

funriser commented on delegates/src/main/java/com/revolut/recyclerkit/delegates/DiffAdapter.kt:
```diff
@@ -25,22 +28,113 @@ import androidx.recyclerview.widget.DiffUtil
+
+        override fun setItems(items: List<ListItem>, recyclerView: RecyclerView) {
+            val (diffResult, newList) = calculateDiff(items)
+            dispatchDiffInternal(diffResult, newList, recyclerView)
```
Please append thread check the same way `RxDiffAdapter` does

funriser commented on delegates/src/main/java/com/revolut/recyclerkit/delegates/DiffAdapter.kt:
```diff
@@ -25,22 +28,113 @@ import androidx.recyclerview.widget.DiffUtil
+        override val items = ArrayList<ListItem>()
+
+        override fun setItems(items: List<ListItem>, recyclerView: RecyclerView) {
+            val (diffResult, newList) = calculateDiff(items)
```
It's unnecessary to return pair of result + newList from `calculateDiff` method. `newList` and `items` properties reference the same list

RKuanysh committed:
Added DifferDelegate inside RxDiffAdapter

funriser commented on delegates/src/main/java/com/revolut/recyclerkit/delegates/DiffAdapter.kt:
```diff
@@ -25,22 +30,146 @@ import androidx.recyclerview.widget.DiffUtil
+            diffResult.dispatchUpdatesTo(adapter)
+        }
+
+        protected fun dispatchDiffInternal(diffResult: DiffUtil.DiffResult, newList: List<ListItem>, recyclerView: RecyclerView) {
```
`SyncDifferStrategy` doesn't use dispatchDiffInternal, therefore ignores `autoScrollToTop` flag

RKuanysh committed:
Fixed an issue

aleksei-tvorogov commented on rxdiffadapter/src/main/java/com/revolut/rxdiffadapter/RxDiffAdapter.kt:
```diff
@@ -47,7 +44,7 @@ open class RxDiffAdapter @Deprecated("Replace with constructor without delegates
     private val autoScrollToTop: Boolean = false,
     private val detectMoves: Boolean = true
-) : AbsRecyclerDelegatesAdapter(delegatesManager) {
+) : DiffAdapter(delegatesManager = delegatesManager, async = async, autoScrollToTop = autoScrollToTop) {
```
DiffAdapter now supports both sync/async mode, therefore we don't need to keep rx-based implementation here. 
Let's deprecate this class and make it type alias for DiffAdapter

aleksei-tvorogov commented on delegates/src/main/java/com/revolut/recyclerkit/delegates/DiffAdapter.kt:
```diff
@@ -25,22 +30,144 @@ import androidx.recyclerview.widget.DiffUtil
+            val rv = recyclerViewRef.get() ?: error("Recycler View not attached")
+            val layoutManager = rv.layoutManager
+
+            val firstVisiblePosition = if (autoScrollToTop && layoutManager is LinearLayoutManager) {
```
There is code duplication for autoScroll logic in SyncDifferStrategy and AsyncSyncDifferStrategy.
- Let's put code into lamda and use it in both cases.
- Let's invoke it always after submitting list to recycler view.  Now it happens in different moments

RKuanysh committed:
Unified scroll to 0 position when isComputingLayout and not.
Using unified extension method to calculate first visible position in recycler view
